src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -108,4 +136,281 @@ protected int getCacheLimit() {
+        // inconvenient because we can't work with ropes directly in Ruby and some MatchData methods are nicely
+        // implemented using the source string data. Likewise, we need to taint objects based on the source string's
+        // taint state. We mustn't allow the source string's contents to change, however, so we must ensure that we have
+        // a private copy of that string. Since the source string would otherwise be a reference to string held outside
+        // the MatchData object, it would be possible for the source string to be modified externally.
+        //
+        // Ex. x = "abc"; x =~ /(.*)/; x.upcase!
+        //
+        // Without a private copy, the MatchData's source could be modified to be upcased when it should remain the
+        // same as when the MatchData was created.
+        @Specialization
+        protected DynamicObject executeMatch(DynamicObject regexp, DynamicObject string, Matcher matcher,
+                int startPos, int range, boolean onlyMatchAtStart,
+                @Cached("createBinaryProfile()") ConditionProfile matchesProfile) {
+            assert RubyGuards.isRubyRegexp(regexp);
+            assert RubyGuards.isRubyString(string);
+
+            if (getContext().getOptions().REGEXP_INSTRUMENT_MATCH) {
+                instrument(regexp, string, onlyMatchAtStart);
+            }
+
+            int match = runMatch(matcher, startPos, range, onlyMatchAtStart);
+
+            if (matchesProfile.profile(match == -1)) {
+                return nil();
+            }
+
+            assert match >= 0;
+
+            final Region region = matcher.getEagerRegion();
+            final DynamicObject dupedString = (DynamicObject) dupNode.call(null, string, "dup");
+            DynamicObject result = allocateNode.allocate(matchDataClass(), Layouts.MATCH_DATA.build(dupedString,
+                    regexp, region, null));
+            return (DynamicObject) taintResultNode.maybeTaint(string, result);
+        }
+
+        @TruffleBoundary
+        private int runMatch(Matcher matcher, int startPos, int range, boolean onlyMatchAtStart) {
+            // Keep status as RUN because MRI has an uninterruptible Regexp engine
+            if (onlyMatchAtStart) {
+                return getContext().getThreadManager().runUntilResultKeepStatus(this,
+                        () -> matcher.matchInterruptible(startPos, range, Option.DEFAULT));
+            } else {
+                return getContext().getThreadManager().runUntilResultKeepStatus(this,
+                        () -> matcher.searchInterruptible(startPos, range, Option.DEFAULT));
+            }
+        }
+
+        @TruffleBoundary
+        protected void instrument(DynamicObject regexp, DynamicObject string, boolean fromStart) {
+            Rope source = Layouts.REGEXP.getSource(regexp);
+            Encoding enc = Layouts.STRING.getRope(string).getEncoding();
```

Comment:
Should this be the Regexp's encoding or the string to match's encoding?

Reply:
The string's, Technically it should really be the encoding calculated when creating the `Matcher` object, and hence the internal joni `Regex` encoding, but it was ugly getting that info out.

Reply:
Isn't it just Layouts.REGEXP.getRegex(regexp).getEncoding() ?